src/main/java/hudson/maven/MavenBuild.java

```diff
@@ -778,116 +778,134 @@ protected Lease decideWorkspace(Node n, WorkspaceList wsl) throws InterruptedExc
+                    margs.add("-Dmaven.repo.local="+localRepo.getRemote());
+
+                String settingsPath = SettingsProvider.getSettingsRemotePath(mms.getSettings(), MavenBuild.this, listener);
+                if (settingsPath != null) {
+                    margs.add("-s").add(settingsPath);
+                }
+
+                margs.add("-f",getModuleRoot().child("pom.xml").getRemote());
+                margs.addTokenized(getProject().getGoals());
+
+                Map<String,String> systemProps = new HashMap<String, String>(envVars);
+                // backward compatibility
+                systemProps.put("hudson.build.number",String.valueOf(getNumber()));
+
+                ProxyImpl proxy;
+                if (mavenBuildInformation.isMaven3OrLater())
+                {
+                    ProxyImpl2 proxy2 = new ProxyImpl2(mms.getLastCompletedBuild(), slistener);
+                    proxy2.setBlockBuildEvents(true);
+                    proxy = proxy2;
+                    builder = new Maven3Builder(createRequest(proxy2,
+                            getProject(), margs.toList(), systemProps,
+                            mavenBuildInformation));
+                }
+                else {
+                    proxy = new ProxyImpl();
+                    builder = new Builder(
+                            listener, proxy,
+                            getProject(), margs.toList(), systemProps);
+                }
                 Result r = process.call(builder);
                 proxy.performArchiving(launcher, listener);
                 normalExit = true;
                 return r;
             } finally {
-                builder.end(launcher);
-                if(normalExit)  process.recycle();
-                else            process.discard();
-    
+                if (builder != null) {
+                    builder.end(launcher);
+                }
+
+                if (process != null) {
+                    if(normalExit)  process.recycle();
+                    else            process.discard();
+                }
+
                 // tear down in reverse order
                 boolean failed=false;
                 for( int i=buildEnvironments.size()-1; i>=0; i-- ) {
-                    if (!buildEnvironments.get(i).tearDown(MavenBuild.this,slistener)) {
+                    final Environment environment = buildEnvironments.get(i);
+                    try {
+                        if (!environment.tearDown(MavenBuild.this,slistener)) {
+                            failed=true;
+                        }
+                    } catch (Throwable inTearDown) {
                         failed=true;
+
+                        // exceptions are ignored to give a chance to all environments to tear down
+                        listener.error("Unable to tear down: " + inTearDown.getMessage());
+                        if (debug) {
+                            inTearDown.printStackTrace(listener.getLogger());
```

Comment:
No such exception is expected, so this should be printed unconditionally. Better yet, send to the logger.